File: Sources/Config/LlmConfig.swift
import Foundation

struct LlmConfig: Codable, Sendable {
    var provider: String = "anthropic"
    var model: String?
    var apiKey: String?
    var baseUrl: String?
    var apiType: String?

    static let providers = ["anthropic", "openai", "groq", "cerebras", "deepseek", "zai", "litellm", "ollama", "custom"]

    static let defaultModels: [String: String] = [
        "anthropic": "claude-sonnet-4-20250514",
        "openai": "gpt-4o",
        "groq": "llama-3.3-70b",
        "cerebras": "llama-3.3-70b",
        "deepseek": "deepseek-chat",
        "zai": "glm-4",
        "litellm": "claude-sonnet-4-6",
        "ollama": "llama3",
        "custom": "model-name",
    ]

    static let envVarKeys: [String: String] = [
        "anthropic": "ANTHROPIC_API_KEY",
        "openai": "OPENAI_API_KEY",
        "groq": "GROQ_API_KEY",
        "cerebras": "CEREBRAS_API_KEY",
        "deepseek": "DEEPSEEK_API_KEY",
        "zai": "ZAI_API_KEY",
        "litellm": "LITELLM_API_KEY",
    ]

    static let defaultBaseUrls: [String: String] = [
        "ollama": "http://localhost:11434",
        "litellm": "https://inference.noizu.com/v1",
    ]

    static let needsApiKey: Set<String> = ["anthropic", "openai", "groq", "cerebras", "deepseek", "zai", "litellm", "custom"]
    static let needsBaseUrl: Set<String> = ["ollama", "litellm", "custom"]

    var effectiveModel: String {
        model ?? LlmConfig.defaultModels[provider] ?? "claude-sonnet-4-20250514"
    }

    var effectiveApiKey: String? {
        if let key = apiKey, !key.isEmpty { return key }
        if let envVar = LlmConfig.envVarKeys[provider] {
            if let val = ProcessInfo.processInfo.environment[envVar], !val.isEmpty { return val }
        }
        if provider == "litellm" {
            return ProcessInfo.processInfo.environment["OPENAI_API_KEY"]
        }
        return nil
    }

    var effectiveBaseUrl: String? {
        if let url = baseUrl, !url.isEmpty { return url }
        return LlmConfig.defaultBaseUrls[provider]
    }
}

File: frontend/src/lib/llm-config.ts
/**
 * LLM provider configuration with autofill support
 * Ported from queue-populator's LlmConfig.swift
 */

export interface LlmProviderConfig {
  name: string;
  label: string;
  defaultModel: string;
  baseUrl: string;
  envKey: string;
  needsApiKey: boolean;
  needsBaseUrl: boolean;
  envFallbacks?: string[];
}

/**
 * Supported LLM providers with their default configurations
 */
export const LLM_PROVIDERS: Record<string, LlmProviderConfig> = {
  anthropic: {
    name: 'anthropic',
    label: 'Anthropic (Claude)',
    defaultModel: 'claude-sonnet-4-6',
    baseUrl: 'https://api.anthropic.com/v1',
    envKey: 'ANTHROPIC_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  openai: {
    name: 'openai',
    label: 'OpenAI',
    defaultModel: 'gpt-4o',
    baseUrl: 'https://api.openai.com/v1',
    envKey: 'OPENAI_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  groq: {
    name: 'groq',
    label: 'Groq',
    defaultModel: 'llama-3.3-70b',
    baseUrl: 'https://api.groq.com/openai/v1',
    envKey: 'GROQ_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  cerebras: {
    name: 'cerebras',
    label: 'Cerebras',
    defaultModel: 'llama-3.3-70b',
    baseUrl: 'https://api.cerebras.ai/v1',
    envKey: 'CEREBRAS_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  deepseek: {
    name: 'deepseek',
    label: 'DeepSeek',
    defaultModel: 'deepseek-chat',
    baseUrl: 'https://api.deepseek.com/v1',
    envKey: 'DEEPSEEK_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  mistral: {
    name: 'mistral',
    label: 'Mistral',
    defaultModel: 'mistral-large',
    baseUrl: 'https://api.mistral.ai/v1',
    envKey: 'MISTRAL_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  gemini: {
    name: 'gemini',
    label: 'Google Gemini',
    defaultModel: 'gemini-2.0-flash-exp',
    baseUrl: 'https://generativelanguage.googleapis.com/v1beta',
    envKey: 'GOOGLE_API_KEY',
    needsApiKey: true,
    needsBaseUrl: false,
  },
  ollama: {
    name: 'ollama',
    label: 'Ollama (Local)',
    defaultModel: 'llama3',
    baseUrl: 'http://localhost:11434',
    envKey: '',
    needsApiKey: false,
    needsBaseUrl: true,
  },
  litellm: {
    name: 'litellm',
    label: 'LiteLLM Proxy',
    defaultModel: 'claude-sonnet-4-6',
    baseUrl: 'https://inference.noizu.com/v1',
    envKey: 'LITELLM_API_KEY',
    needsApiKey: true,
    needsBaseUrl: true,
    envFallbacks: ['OPENAI_API_KEY'],
  },
  custom: {
    name: 'custom',
    label: 'Custom (OpenAI-compatible)',
    defaultModel: 'model-name',
    baseUrl: 'https://api.example.com/v1',
    envKey: '',
    needsApiKey: false,
    needsBaseUrl: true,
  },
};

/**
 * Get provider configuration by name
 */
export function getProviderConfig(provider: string): LlmProviderConfig | undefined {
  return LLM_PROVIDERS[provider];
}

/**
 * Get a list of provider names
 */
export function getProviderNames(): string[] {
  return Object.keys(LLM_PROVIDERS);
}

/**
 * Check if provider needs API key
 */
export function providerNeedsApiKey(provider: string): boolean {
  return LLM_PROVIDERS[provider]?.needsApiKey ?? false;
}

/**
 * Check if provider needs custom base URL
 */
export function providerNeedsBaseUrl(provider: string): boolean {
  return LLM_PROVIDERS[provider]?.needsBaseUrl ?? false;
}

/**
 * Get default model for provider
 */
export function getDefaultModel(provider: string): string {
  return LLM_PROVIDERS[provider]?.defaultModel ?? 'model-name';
}

/**
 * Get default base URL for provider
 */
export function getDefaultBaseUrl(provider: string): string {
  return LLM_PROVIDERS[provider]?.baseUrl ?? 'https://api.example.com/v1';
}

/**
 * Get environment variable name for provider's API key
 */
export function getProviderEnvKey(provider: string): string {
  return LLM_PROVIDERS[provider]?.envKey ?? '';
}

/**
 * Get fallback environment variable names for provider
 */
export function getProviderEnvFallbacks(provider: string): string[] {
  return LLM_PROVIDERS[provider]?.envFallbacks ?? [];
}

/**
 * Provider label for display
 */
export function getProviderLabel(provider: string): string {
  return LLM_PROVIDERS[provider]?.label ?? provider;
}